uvicorn/protocols/http/httparse_impl.py

```diff
@@ -0,0 +1,616 @@
+        self.scheme = "https" if is_ssl(transport) else "http"
+
+        if self.logger.level <= TRACE_LOG_LEVEL:
+            prefix = "%s:%d - " % self.client if self.client else ""
+            self.logger.log(TRACE_LOG_LEVEL, "%sHTTP connection made", prefix)
+
+    def connection_lost(self, exc: Optional[Exception]) -> None:
+        self.connections.discard(self)
+
+        if self.logger.level <= TRACE_LOG_LEVEL:
+            prefix = "%s:%d - " % self.client if self.client else ""
+            self.logger.log(TRACE_LOG_LEVEL, "%sHTTP connection lost", prefix)
+
+        if self.cycle and not self.cycle.response_complete:
+            self.cycle.disconnected = True
+        if self.cycle is not None:
+            self.cycle.message_event.set()
+        if self.flow is not None:
+            self.flow.resume_writing()
+        if exc is None:
+            self.transport.close()
+            self._unset_keepalive_if_required()
+
+        self.parser = httparse.RequestParser()
+
+    def _unset_keepalive_if_required(self) -> None:
+        if self.timeout_keep_alive_task is not None:
+            self.timeout_keep_alive_task.cancel()
+            self.timeout_keep_alive_task = None
+
+    def _get_upgrade(self) -> Optional[bytes]:
+        connection = []
+        upgrade = None
+        for name, value in self.headers:
+            if name == b"connection":
+                connection = [token.lower().strip() for token in value.split(b",")]
+            if name == b"upgrade":
+                upgrade = value.lower()
+        if b"upgrade" in connection:
+            return upgrade
+        return None
+
+    def _should_upgrade_to_ws(self) -> bool:
+        if self.ws_protocol_class is None:
+            if self.config.ws == "auto":
+                msg = "Unsupported upgrade request."
+                self.logger.warning(msg)
+                msg = "No supported WebSocket library detected. Please use 'pip install uvicorn[standard]', or install 'websockets' or 'wsproto' manually."  # noqa: E501
+                self.logger.warning(msg)
+            return False
+        return True
+
+    def data_received(self, data: bytes) -> None:
+        self._unset_keepalive_if_required()
+
+        self._buffer += data
+
+        if self._parsed is None:
+            self._attempt_to_parse_request()
+        else:
+            self._consume_body()
+
+    def eof_received(self) -> None:
+        if self._parsed is None:
+            msg = "Invalid HTTP request received."
+            self.logger.warning(msg)
+            self.send_400_response(msg)
+            return
+
+        if self.cycle is not None:
+            self.cycle.more_body = False
+            self.cycle.message_event.set()
+
+        self._parsed = None
+
+    def _attempt_to_parse_request(self) -> None:
+        try:
+            parsed = self.parser.parse(self._buffer)
+        except httparse._httparse.ParsingError:
+            return
```

Comment:
Aren't we supposing to do something if there's an error when parsing the headers? I guess this is not partial, as partial results is parsed being `None`, right?

Reply:
Yes, a partial (non-invalid) request will lead to `None`.

Maybe I should add a commit here. We don't do anything special here, because that leaves `_parsed` as `None`. Then, when `eof_received()` is called, it sees that `_parsed is None`, we surely means we didn't get any valid request — and that's where we send the 400 Bad Request response. This also covers the case when no data was sent at all, or the client sent a partial request and never finished it. So I think it's the most complete way to do things. But I agree this "just `return`" line here prompts questions... I'll defo add a comment.